File: code/server/jvm/im-domain/src/main/kotlin/im/domain/wrappers/name/NameValidationError.kt
package im.domain.wrappers.name

sealed class NameValidationError(
    private val defaultMessage: String = "Name is invalid",
) {
    data object Blank : NameValidationError("Name cannot be blank")

    data class InvalidLength(
        val min: Int,
        val max: Int,
    ) : NameValidationError("Name must be between $min and $max characters")

    fun toErrorMessage(): String = defaultMessage
}

fun List<NameValidationError>.toErrorMessage(): String = joinToString("\n") { it.toErrorMessage() }

File: code/client/web/src/Domain/wrappers/name/NameValidationError.ts
export abstract class NameValidationError {
    private readonly defaultMessage: string;

    protected constructor(defaultMessage: string = 'Name is invalid') {
        this.defaultMessage = defaultMessage;
    }

    static Blank = new (class extends NameValidationError {
        constructor() {
            super('Name cannot be blank');
        }
    })();

    static InvalidLength(min: number, max: number) {
        return new (class extends NameValidationError {
            constructor() {
                super(`Name must be between ${min} and ${max} characters`);
            }
        })();
    }

    toErrorMessage(): string {
        return this.defaultMessage;
    }
}
